**Conservation management strategy impacts inbreeding and mutation load in scimitar-horned oryx**

**Summary**
-------------
This repository contains the scripts used for downstream analyses of population structure, ROH and deleterious mutations. Please see github repository [SHO_reseq_2022](https://github.com/elhumble/SHO_reseq_2022) for prior steps including alignment, SNP calling and variant annotation.

**Code structure**
-------------

*Relatedness and population structure*  
**1.0_IBD.R:**  
**1.1_ngsadmix_out.R:**  
**1.2_pcangsd_out.R:**  

*ROH analysis*  
**2.0_ROH_plink.R:**  
**2.1_ROH_bcftools.R:**  
**2.2_ROH_plink_supp:**  

*Outgroup depth*  
**3.0_outgroup_depth.R:**  

*Mutation load analysis*  
**4.0_load_snpeff.R:**  
**4.1_load_vep.R:**  
**4.2_load_snpeff_supp.R:**  

*Ne and fixed loci*  
**5.0_gone.R:**  
**6.0_fixed.R:**  

*Coverage assessment*  
**0.1_vcf_cov.R:**  
**6.1_coverage.R:**  

*File creation*  
**0.0_chrom_map.R:**  

**Data**
-------------
Data files can be generated using scripts from the [first part](https://github.com/elhumble/SHO_reseq_2022) of the pipeline.  

###### Please feel free to [get in touch](mailto:emily.humble@ed.ac.uk) if you have any questions.
